ethereum/contracts/ShutdownSwitch.sol

```diff
@@ -0,0 +1,134 @@
+    mapping(address => bool) private disabledVotes;
+
+    bool private enabled = true;
+
+    /// @dev Returns the number of votes required to disable transfers.
+    function requiredVotesToDisable() public view returns (uint16) {
+        return computeRequiredVotesToDisable(getCurrentGuardianSet().keys.length);
+    }
+
+    /// @dev Returns the current number of votes to disable transfers.
+    function numVotesToDisable() public view returns (uint16) {
+        return computeNumVotesDisabled(getCurrentGuardianSet());
+    }
+
+    /// @dev returns the current shutdown status, where true means transfers are enabled.
+    function enabledFlag() public view returns (bool) {
+        return enabled;
+    }
+
+    /// @dev Event published whenenver a valid guardian votes.
+    event ShutdownVoteCast(address indexed voter, bool votedToEnable, uint16 numVotesToDisable, bool enableFlag);
+
+    /// @dev Event published whenever the shutdown status changes from enabled to disabled, or vice versa.
+    event ShutdownStatusChanged(bool enabledFlag, uint16 numVotesToDisable);
+
+    /// @dev Function that must be implemented by contracts inheriting from this one.
+    function getCurrentGuardianSet() public virtual view returns (Structs.GuardianSet memory);
+
+    /// @dev The number of disable votes required to disable transfers (assuming there are at least that many guardians).
+    uint16 constant REQUIRED_NO_VOTES = 3;
+
+    /// @dev Function that should be called from setup() on new contract deployments, or initialize() on the initial upgrade to deploy this feature.
+    function setUpShutdownSwitch() internal {
+        enabled = true;
+    }
+
+    /// @dev modifier used to block transfers when shutdown.
+    modifier isEnabled {
+        require(enabledFlag(), "transfers are temporarily disabled");
+        _;
+    }
+
+    /// @dev This is the function that allows guardians to vote, and determines the resulting shutdown status.
+    function castShutdownVote(bool _enabled) public {
+        // We always use the current guardian set.
+        Structs.GuardianSet memory gs = getCurrentGuardianSet();
+
+        // Only currently active guardians are allowed to vote.
+        require(isRegisteredVoter(msg.sender, gs), "you are not a registered voter");
+
+        // Only votes to disable are maintained in the map.
+        if (_enabled) {
+            delete disabledVotes[msg.sender];
+        } else {
+            disabledVotes[msg.sender] = true;
+        }
+
+        // Update the number of votes to disable.
+        uint16 votesToDisable = computeNumVotesDisabled(gs);
+
+        // Determine the new shutdown status and generate the appropriate events.
+        bool newEnabledFlag = (votesToDisable < computeRequiredVotesToDisable(gs.keys.length));
+        emit ShutdownSwitch.ShutdownVoteCast(msg.sender, _enabled, votesToDisable, newEnabledFlag);
```

Comment:
So we emit an event even when the voter votes twice in the same direction. Is that intentional?

Reply:
It's not a concern, but it will emit repeated events, and just wanted to check if that's OK

Reply:
Before the rework, I only generated the event when something changed, but as I reworked it, I decided to always generate the event when a valid voter casts a vote, even if it doesn't change anything. It might be worthwhile to know when guardians are voting. . .

Reply:
So if there are no objections, I am going to leave it like it is.